### ISSUE TITLE
multiple states and scheduling

### ISSUE THREAD

**Issue by jfouche** [bevyengine/bevy#8234]
## What you did

I'm starting learning new states in Bevy. I tried the following code, assuming that when using 2 states, I could manage both when defining set and scheduling :

```rust 
    use bevy::prelude::*;

    #[derive(States, Debug, Clone, Copy, Eq, PartialEq, Hash, Default)]
    enum AppState {
        #[default]
        MainMenu,
        InGame,
    }

    #[derive(States, Debug, Clone, Copy, Eq, PartialEq, Hash, Default)]
    enum SimulationState {
        Running,
        #[default]
        Paused,
    }

    #[test]
    fn test_multiple_states() {
        let mut app = App::new();
        app.add_state::<AppState>()
            .add_state::<SimulationState>()
            .add_system(
                spawn_pause_menu
                    .in_set(OnUpdate(AppState::InGame))
                    .in_schedule(OnEnter(SimulationState::Paused)),
            )
            .run();

        fn spawn_pause_menu(mut _commands: Commands) {
            panic!("Pause menu spawned");
        }
    }
```

## What went wrong

Here, the test failed, because the pause menu was spawn but the AppState is not InGame, but MainMenu. I suppose that the `spawn_pause_menu` is called because it entered `SimulationState::Paused`, not checking the `in_set(OnUpdate(AppState::InGame))`. Is that expected? If the answer is yes, I would like an explanation on how should I manage this case.

**Comment by alice-i-cecile** [comment]
Use `in_state(AppState::InGame)` . `OnEnter` only works in the main schedule.

**Comment by JoJoJet** [comment]
I believe this is resolved by #8260, since removing the `OnUpdate` set also removes this footgun.
